### DISCUSSION TITLE
Add ability to cancel batched job spawning

### DISCUSSION THREAD

**Discussion by bencroker** [craftcms/cms#14986] [Ideas] [answered]
The way that batched queue jobs currently work, cancelling a batched job has little or no effect, as the next job is spawned within the `execute()` method.

https://github.com/craftcms/cms/blob/6a13d279272e81fec9cd919aba58c497a2febd78/src/queue/BaseBatchedJob.php#L153-L160

The only way to truly cancel a batched job is to cancel the queue job and kill the process running it. 

This has resulted in several support requests in which people have been confused about why cancelling long-running, batched queue jobs has no effect. For example:

> One thing I am noticing during my testing (currently debugging an access issue preventing the cache from generating in these queue jobs) is that subsequent jobs aren't added to the queue until the current batch has completed. So trying to cancel everything is pretty tedious: cancel the current batch, hit OK, refresh, repeat. Is it possible to have them all added at once so they could be released at the same time?

Source: https://github.com/putyourlightson/craft-blitz/issues/537#issuecomment-2008482896

> I think what I'm looking for in regards to the UX is less batching and more having them all added at once but as separate jobs. Using those screenshots as examples, it would add 24 separate jobs all at once while still only running one at a time (assuming I only have one process running). But I'd also be able to release them all at once instead of individually. 

Source: https://github.com/putyourlightson/craft-blitz/issues/537#issuecomment-2009873023

I’m not sure if and how you’d consider addressing this. One option might be to add a check to the `execute()` method that ensures that the queue job still exists in the DB before spawning a new one. Any thoughts on this?

**Comment by russback** [comment]
+1 for this - for large sites this can be problematic. A preflight check as @bencroker suggests sounds like a good idea, or something else that checks whether that 'Release all Jobs' button has been hit since the initial queue request was made.

**Comment by brandonkelly** [comment, marked as answer]
Thanks for pointing that out! As of Craft 4.10 and 5.2, batched jobs will start ensuring they are still reserved before moving onto the next step or spawning an additional batch job. (6505856bbfd35c3afa48be084fd91f9ffb8d5672)

**Reply by bencroker** [reply]
Great, thanks!

**Reply by brandonkelly** [reply]
Craft 4.10.0-beta.1 and 5.2.0-beta.1 are out with that change.
